### PR TITLE
Add isort (for sorting imports)

### PR DESCRIPTION
**Pull request by mdesmet**
Introduces isort (sorting imports)

- Added isort as "test" dependencies
- Run `pre-commit run --all-files` to standardize formatting and import sorting.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -16,7 +16,7 @@
 import re
 import textwrap
 
-from setuptools import setup, find_packages
+from setuptools import find_packages, setup
```
Comment (nineinchnick) [inline setup.py:19]: Formatting changes should be in a separate commit.

**Review by nineinchnick — APPROVED** [review]

File: setup.py (hunk 1 above)
Comment (mdesmet) [reply setup.py:19]: A separate commit would fail the build if ran by itself.

Comment (nineinchnick) [reply setup.py:19]: Why? Make it the first commit.

File: README.md [hunk 2]
```diff
@@ -424,6 +424,7 @@ We recommend that you use Python3's `venv` for development:
 $ python3 -m venv .venv
 $ . .venv/bin/activate
 $ pip install -e '.[tests]'
+$ pre-commit install
```
Comment (hashhar) [inline README.md:427]: this needs pre-commit cli to be installed (which isn't by default AFAIK).

File: README.md [hunk 3]
```diff
@@ -441,6 +442,17 @@ When the code is ready, submit a Pull Request.
 See also Trino's [guidelines](https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md).
 Most of them also apply to code in trino-python-client.
 
+### `pre-commit` checks
+
+Code is automatically checked on commit by a [pre-commit](https://pre-commit.com/) git hook.
+
+Following checks are performed:
+
+- [`flake8`](https://flake8.pycqa.org/en/latest/) for code linting
```
Comment (hashhar) [inline README.md:451]: linting should run on CI for now. We don't want to prevent from submitting PRs or pushing changes just when experimenting or when one needs help to fix whatever is being complained about.

File: README.md [hunk 4]
```diff
@@ -441,6 +442,17 @@ When the code is ready, submit a Pull Request.
...
+- [`black`](https://github.com/psf/black) for code formatting
+- [`isort`](https://pycqa.github.io/isort/) for sorting imports
+- [`mypy`](https://mypy.readthedocs.io/en/stable/) for static type checking
```
Comment (hashhar) [inline README.md:454]: `mypy` on CI only for now as well.

File: tests/integration/conftest.py [hunk 5]
```diff
@@ -64,13 +64,7 @@ def start_trino(image_tag=None):
...
 
 def wait_for_trino_workers(host, port, timeout=180):
-    request = TrinoRequest(
-        host=host,
-        port=port,
-        client_session=ClientSession(
-            user="test_fixture"
-        )
-    )
+    request = TrinoRequest(host=host, port=port, client_session=ClientSession(user="test_fixture"))
```
Comment (hashhar) [inline tests/integration/conftest.py:67]: arguably less readable than before.

File: tests/integration/test_dbapi_integration.py [hunk 6]
```diff
@@ -153,7 +151,9 @@ def test_execute_many_without_params(trino_connection):
     cur = trino_connection.cursor()
     cur.execute("CREATE TABLE memory.default.test_execute_many_without_param (value varchar)")
     cur.fetchall()
-    cur.executemany("INSERT INTO memory.default.test_execute_many_without_param (value) VALUES (?)", [])
+    cur.executemany(
+        "INSERT INTO memory.default.test_execute_many_without_param (value) VALUES (?)", []
```
Comment (hashhar) [inline tests/integration/test_dbapi_integration.py:155]: Split query across lines instead of moving argument to newline

File: tests/integration/test_dbapi_integration.py [hunk 7]
```diff
@@ -195,43 +194,45 @@ def test_experimental_python_types_with_connection_and_cursor(
 
     cur = connection.cursor(experimental_python_types=cursor_experimental_python_types)
 
-    cur.execute("""
+    cur.execute(
+        """
     SELECT
```
Comment (hashhar) [inline tests/integration/test_dbapi_integration.py:199]: Move SELECT to same line as `"""` then.

File: tests/integration/test_dbapi_integration.py [hunk 8]
```diff
@@ -877,8 +886,9 @@ def test_transaction_multiple(trino_connection_with_transaction):
     assert len(rows2) == 1000
 
 
-@pytest.mark.skipif(trino_version() == '351', reason="Autocommit behaves "
-                                                     "differently in older Trino versions")
+@pytest.mark.skipif(
+    trino_version() == "351", reason="Autocommit behaves " "differently in older Trino versions"
```
Comment (hashhar) [inline tests/integration/test_dbapi_integration.py:890]:
```suggestion
    trino_version() == "351", reason="Autocommit behaves differently in older Trino versions"
```

File: tests/integration/test_dbapi_integration.py [hunk 9]
```diff
@@ -887,17 +897,18 @@ def test_transaction_autocommit(trino_connection_in_autocommit):
             """
             CREATE TABLE memory.default.nation
             AS SELECT * from tpch.tiny.nation
-            """)
+            """
+        )
```
Comment (hashhar) [inline tests/integration/test_dbapi_integration.py:901]: why?

File: tests/integration/test_sqlalchemy_integration.py [hunk 10]
```diff
@@ -112,143 +118,159 @@ def test_insert(trino_connection):
...
-            customers.c.mktsegment == 'HOUSEHOLD'
-        ),
-        not_(customers.c.acctbal < 0)))
+    customers = sqla.Table("customer", metadata, schema="tiny", autoload_with=conn)
+    query = sqla.select(customers).where(
+        and_(
+            customers.c.name.like("%12%"),
+            customers.c.nationkey == 15,
+            or_(customers.c.mktsegment == "AUTOMOBILE", customers.c.mktsegment == "HOUSEHOLD"),
+            not_(customers.c.acctbal < 0),
```
Comment (hashhar) [inline tests/integration/test_sqlalchemy_integration.py:185]: Does black reformat this away?

```suggestion
        and_(
            customers.c.name.like("%12%"),
            customers.c.nationkey == 15,
        or_(customers.c.mktsegment == "AUTOMOBILE", customers.c.mktsegment == "HOUSEHOLD"),
        not_(customers.c.acctbal < 0),
```

File: tests/unit/conftest.py [hunk 11]
```diff
@@ -194,8 +195,7 @@ def sample_get_error_response_data():
             "errorType": "USER_ERROR",
             "failureInfo": {
                 "errorLocation": {"columnNumber": 15, "lineNumber": 1},
-                "message": "line 1:15: Schema must be specified "
-                "when session schema is not set",
+                "message": "line 1:15: Schema must be specified " "when session schema is not set",
```
Comment (hashhar) [inline tests/unit/conftest.py:198]:
```suggestion
                "message": "line 1:15: Schema must be specified when session schema is not set",
```

File: tests/unit/oauth_test_utils.py [hunk 12]
```diff
@@ -45,9 +45,15 @@ def __call__(self, request, uri, response_headers):
...
             return [200, response_headers, json.dumps(self.sample_post_response_data)]
-        return [401, {'Www-Authenticate': f'Bearer x_redirect_server="{self.redirect_server}", '
-                                          f'x_token_server="{self.token_server}"',
-                      'Basic realm': '"Trino"'}, ""]
+        return [
+            401,
+            {
+                "Www-Authenticate": f'Bearer x_redirect_server="{self.redirect_server}", '
+                f'x_token_server="{self.token_server}"',
+                "Basic realm": '"Trino"',
```
Comment (hashhar) [inline tests/unit/oauth_test_utils.py:53]: More readable and "safer" to have entire token on one line

File: tests/unit/sqlalchemy/test_compiler.py [hunk 13]
```diff
@@ -47,7 +35,10 @@ def dialect():
 def test_limit_offset(dialect):
     statement = select(table).limit(10).offset(0)
     query = statement.compile(dialect=dialect)
-    assert str(query) == 'SELECT "table".id, "table".name \nFROM "table"\nOFFSET :param_1\nLIMIT :param_2'
+    assert (
+        str(query)
+        == 'SELECT "table".id, "table".name \nFROM "table"\nOFFSET :param_1\nLIMIT :param_2'
+    )
```
Comment (hashhar) [inline tests/unit/sqlalchemy/test_compiler.py:41]: questionable redability

File: README.md (hunk 3 above)
Comment (mdesmet) [reply README.md:451]: That can be disabled by `git commit . -m 'quick fix' --no-verify`. We can add this to the docs as well. 

Note that in general only mypy might stop you. black and isort just autoformat your files.

IMHO in general it will be better to always run it on commit to prevent contributors from the experience of failing the build for easily fixable problems.

WDYT?

File: .pre-commit-config.yaml [hunk 14]
```diff
@@ -13,3 +13,16 @@ repos:
         additional_dependencies:
           - "types-pytz"
           - "types-requests"
+
+  - repo: https://github.com/psf/black
+    rev: 22.3.0
+    hooks:
+      - id: black
+        args:
+          - "--line-length=99"
```
Comment (hashhar) [inline .pre-commit-config.yaml:22]: enforcing a specific line length can hinder readability - line length limits don't provide any benefits IMO.
Let's set to `120` since that's still reasonable.

FWIW here's existing max line lengths.

```
(.venv) hashhar@Ashhars-MacBook-Pro trino-python-client % git ls-files | xargs -n1 gwc -L | sort -n
0 tests/__init__.py
0 tests/integration/__init__.py
0 tests/unit/__init__.py
0 tests/unit/sqlalchemy/__init__.py
30 tox.ini
54 .pre-commit-config.yaml
74 tests/unit/test_transaction.py
74 trino/constants.py
74 trino/logging.py
74 trino/sqlalchemy/__init__.py
75 trino/__init__.py
77 LICENSE
79 tests/unit/test_http.py
80 tests/unit/sqlalchemy/conftest.py
81 trino/exceptions.py
87 tests/integration/conftest.py
87 tests/unit/sqlalchemy/test_datatype_split.py
87 trino/transaction.py
95 .gitignore
95 trino/sqlalchemy/compiler.py
96 tests/unit/oauth_test_utils.py
97 setup.cfg
98 setup.py
101 trino/sqlalchemy/datatype.py
102 tests/unit/sqlalchemy/test_datatype_parse.py
106 tests/unit/sqlalchemy/test_compiler.py
109 tests/integration/test_dbapi_integration.py
109 tests/unit/sqlalchemy/test_dialect.py
110 tests/integration/test_sqlalchemy_integration.py
116 trino/sqlalchemy/error.py
117 trino/auth.py
117 trino/client.py
118 tests/unit/test_dbapi.py
118 trino/sqlalchemy/dialect.py
120 trino/dbapi.py
131 .github/workflows/ci.yml
184 tests/unit/conftest.py
192 tests/unit/test_client.py
438 README.md
```

File: README.md (hunk 3 above)
Comment (hashhar) [reply README.md:451]: I agreed with black and isort. I just meant to keep flake8 and mypy outside of pre-commit for now. We can iterate.

File: tests/integration/conftest.py (hunk 5 above)
Comment (mdesmet) [reply tests/integration/conftest.py:67]: This seems linked to the setting of the line length. It seems black wants to shove as much on the same line if permitted by the line length. No other knobs available. black is fairly opinionated.

File: tests/integration/test_dbapi_integration.py (hunk 6 above)
Comment (mdesmet) [reply tests/integration/test_dbapi_integration.py:155]: Increasing the line length fixed those occurences.

File: tests/integration/test_sqlalchemy_integration.py (hunk 10 above)
Comment (mdesmet) [reply tests/integration/test_sqlalchemy_integration.py:185]: Actually the `_or(...)` is  an argument to the `_and` function, so this seems correct.

File: tests/unit/oauth_test_utils.py (hunk 12 above)
Comment (mdesmet) [reply tests/unit/oauth_test_utils.py:53]: Fixed by increasing line length.

File: tests/unit/sqlalchemy/test_compiler.py (hunk 13 above)
Comment (mdesmet) [reply tests/unit/sqlalchemy/test_compiler.py:41]: A multi line string would be more fitting here. However that triggers flake8 as it seems flake8 doesn't allow spaces at EOL, even within multi line strings.

File: tests/integration/test_dbapi_integration.py (hunk 9 above)
Comment (mdesmet) [reply tests/integration/test_dbapi_integration.py:895]: Changed to one liner with single quotes

File: .pre-commit-config.yaml [hunk 15]
```diff
@@ -4,6 +4,7 @@ repos:
     hooks:
       - id: "flake8"
         name: "Python: analysis"
+        stages: [manual]
```
Comment (hovaesco) [inline .pre-commit-config.yaml:7]: `all` might be more descriptive, since it runs all checks?

**Review by hovaesco — APPROVED** [review]
LGTM % black complains in CI

File: .pre-commit-config.yaml (hunk 15 above)
Comment (mdesmet) [reply .pre-commit-config.yaml:7]: See https://pre-commit.com/#confining-hooks-to-run-at-certain-stages

File: tests/integration/conftest.py (hunk 5 above)
Comment (john-bodley) [reply tests/integration/conftest.py:67]: @mdesmet if you add a trailing comma after `client_session=ClientSession(user="test_fixture"`, Black will format the code is a similar way to how it was formatted previously.

File: .pre-commit-config.yaml [hunk 16]
```diff
@@ -13,3 +14,17 @@ repos:
...
+    hooks:
+      - id: black
+        args:
+          - "--line-length=120"
+
+  - repo: https://github.com/pycqa/isort
+    rev: 5.6.4
+    hooks:
+      - id: isort
+        args: [ "--profile", "black", "--filter-files" ]
```
Comment (john-bodley) [inline .pre-commit-config.yaml:30]: Why are these args required? I've used the `isort` pre-commit successfully with a number of projects which haven't required any additional arguments.

File: .pre-commit-config.yaml [hunk 17]
```diff
@@ -13,3 +14,17 @@ repos:
...
           - "types-pytz"
           - "types-requests"
+        stages: [manual]
+
+  - repo: https://github.com/psf/black
+    rev: 22.3.0
+    hooks:
+      - id: black
+        args:
+          - "--line-length=120"
```
Comment (john-bodley) [inline .pre-commit-config.yaml:24]: Any reason to override the default 88 character line length? I realize it's somewhat arbitrary what the line length is if you have an auto-formatter, but as far as I can tell most packages adhere to the default.

Comment (hashhar) [reply .pre-commit-config.yaml:24]: Because line length doesn't correlate to any reasonable thing. It's an arbitrary rule to be followed and if you'd take a look at the reformats black did with shorter line lengths have questionable readability.

File: .pre-commit-config.yaml (hunk 16 above)
Comment (mdesmet) [reply .pre-commit-config.yaml:30]: The `filter-files` can be omitted. I think it's just there to not change any unmodified files, even if explicitly asked.

The `--profile "black"` actually formats the imports differently

with the black profile

```
from trino.exceptions import (
    DatabaseError,
    DataError,
    Error,
    IntegrityError,
    InterfaceError,
    InternalError,
    NotSupportedError,
    OperationalError,
    ProgrammingError,
    Warning,
)
```

without the profile

```
from trino.exceptions import (DatabaseError, DataError, Error, IntegrityError,
                              InterfaceError, InternalError, NotSupportedError,
                              OperationalError, ProgrammingError, Warning)
```

I personally like the black profile more as it is faster to scan.

**Comment by mdesmet** [comment]
Some flakyness in the tests, might be related to not having merged #220 but not totally sure.

**Comment by hashhar** [comment]
@mdesmet See also https://github.com/trinodb/trino-python-client/commit/cffd2b232d9298321c34430f0ac829f6b6ebea8f#r83714277 - not sure if that's the cause.

**Comment by mdesmet** [comment]
@hashhar: As discussed I have removed black from this PR and only kept isort. PTAL

File: .flake8 [hunk 18]
```diff
@@ -0,0 +1,12 @@
+[flake8]
```
Comment (hashhar) [inline .flake8:1]: intended to be part of this PR or leftover? I think leftover?

File: .github/DEVELOPMENT.md [hunk 19]
```diff
@@ -25,6 +25,18 @@ With `-e` passed to `pip install` above pip can reference the code you are
 modifying in the *virtual env*. That way, you do not need to run `pip install`
 again to make your changes applied to the *virtual env*.
 
+### `pre-commit` checks
```
Comment (hashhar) [inline .github/DEVELOPMENT.md:28]: What exact checks are run will change over time - doesn't seem as important.

Maybe we'd want to instead mention about `pre-commit install` at `Code style` header which already tells how to run these checks manually.

Comment (mdesmet) [reply .github/DEVELOPMENT.md]: Done

File: .flake8 (hunk 18 above)
Comment (mdesmet) [reply .flake8]: Removed

**Comment by mdesmet** [comment]
@hashhar: PTAL

**Review by hashhar — APPROVED** [review]
